Mage/src/main/java/mage/abilities/keyword/ForetellAbility.java

```diff
@@ -348,53 +349,51 @@ public ActivationStatus canActivate(UUID playerId, Game game) {
-                        } else if (((SplitCard) card).getRightHalfCard().getName().equals(abilityName)) {
-                            return ((SplitCard) card).getRightHalfCard().getSpellAbility().canActivate(playerId, game);
-                        }
-                    } else if (card instanceof ModalDoubleFacedCard) {
-                        if (((ModalDoubleFacedCard) card).getLeftHalfCard().getName().equals(abilityName)) {
-                            return ((ModalDoubleFacedCard) card).getLeftHalfCard().getSpellAbility().canActivate(playerId, game);
-                        } else if (((ModalDoubleFacedCard) card).getRightHalfCard().getName().equals(abilityName)) {
-                            return ((ModalDoubleFacedCard) card).getRightHalfCard().getSpellAbility().canActivate(playerId, game);
+                    if (mainCard instanceof CardWithHalves) {
+                        if (((CardWithHalves) mainCard).getLeftHalfCard().getName().equals(faceCardName)) {
+                            return canActivateUsingForetell(((CardWithHalves) mainCard).getLeftHalfCard().getSpellAbility(), playerId, game);
+                        } else if (((CardWithHalves) mainCard).getRightHalfCard().getName().equals(faceCardName)) {
+                            return canActivateUsingForetell(((CardWithHalves) mainCard).getRightHalfCard().getSpellAbility(), playerId, game);
                         }
                     } else if (card instanceof AdventureCard) {
-                        if (card.getMainCard().getName().equals(abilityName)) {
-                            return card.getMainCard().getSpellAbility().canActivate(playerId, game);
-                        } else if (((AdventureCard) card).getSpellCard().getName().equals(abilityName)) {
-                            return ((AdventureCard) card).getSpellCard().getSpellAbility().canActivate(playerId, game);
+                        if (card.getMainCard().getName().equals(faceCardName)) {
+                            return canActivateUsingForetell(card.getMainCard().getSpellAbility(), playerId, game);
+                        } else if (((AdventureCard) card).getSpellCard().getName().equals(faceCardName)) {
+                            return canActivateUsingForetell(((AdventureCard) card).getSpellCard().getSpellAbility(), playerId, game);
                         }
                     }
-                    return card.getSpellAbility().canActivate(playerId, game);
+                    return canActivateUsingForetell(card.getSpellAbility(), playerId, game);
                 }
             }
             return ActivationStatus.getFalse();
         }
 
+        private ActivationStatus canActivateUsingForetell(SpellAbility spellAbility, UUID playerId, Game game) {
+            return spellAbility.copy().setSpellAbilityCastMode(SpellAbilityCastMode.FORETELL).canActivate(playerId, game);
```

Comment:
wtf

Reply:
Do you understand the problem this is trying to solve?

Reply:
Yes, I understand -- you introduced new workaround with injecting global identifier in abilities, it doesn't work with game engine and that's second workaround required to support it.